include/cpp2regex.h2

```diff
@@ -130,6 +134,13 @@ match_context: <CharT, Iter, max_groups: int> type =
     //
     fail: (in this)            match_return<Iter>(false, end);
     pass: (in this, cur: Iter) match_return<Iter>(true, cur);
+
+    //  Private functions
+    //
+    private get_group_idx: (in this, group) -> size_t = {
```

Comment:
I thought "since this is always used to access an element of `groups`, can we make a function that handles that".  I ended up with this:
```
    private get_group: (inout this, group) -> forward match_group<Iter> = {
        assert(group >= 0);
        return groups[group as size_t];
    }
```
and then used like this:
```
    get_group(group).start = pos;
```

I discovered that in order to get a modifiable `match_group` back, the function needed `inout this` and a return type of `forward match_group<Iter>`.

@hsutter  That's a little ugly.  I realize that you want to eliminate references as anything but input parameters to functions, but I hadn't really thought about what that would mean for trying to write functions like this.   Is your expectation that we'd do this?

```
    private get_group: (inout this, group) -> *match_group<Iter> = {
        assert(group >= 0);
        return groups[group as size_t]&;
    }
```
```
        get_group(group)*.start = pos;
```

Reply:
> to get a modifiable `match_group` back, the function needed `inout this` and a return type of `forward match_group<Iter>`.

In general, if you're going to return a `forward`ed modifiable parameter, that parameter needs to be modifiable == `inout`:

```cpp
f: (inout x: int) -> forward int = x;

main: () = {
    xx := 42;
    f(xx) = 43;  // ok, changes xx to 43
    std::cout << xx;  // prints 43
}
```

Does this example answer what you were asking?

Reply:
> Does this example answer what you were asking?

No, I'm talking about writing the Cpp2 equivalent of `T& vector<T>::operator[](size_t index);`.

Reply:
> I thought "since this is always used to access an element of `groups`, can we make a function that handles that".

Yes, that was exactly my thinking at first too! 😃 

Because the `match_context` type has both const and non-const member functions that would need to call this new helper function, it meant I had to write two overloads of the same getter function that returns either a const-ref or a mutable-ref to the element in the groups array, with duplicate implementations.

```cpp
private get_group: (in this, group_idx) -> forward match_group<Iter> = {
    assert(group_idx >= 0);
    return groups[group_idx as size_t];
}

private get_group: (inout this, group_idx) -> forward match_group<Iter> = {
    assert(group_idx >= 0);
    return groups[group_idx as size_t];
}
```
These lower to (slightly edited for clarity):
```cpp
auto get_group(auto const& group_idx) const& -> match_group<Iter> const&;
auto get_group(auto const& group_idx) &      -> match_group<Iter>&;
```

I'm not sure if this is the intended way to write functions that return const-ref or mutable-ref but, just like you found Greg, that was the only way I could achieve it.

(In the real version of the above, there's already a public `get_group` function that returns the element by value, so I chose a different name.)

I didn't like this duplication so I went with the `get_group_idx` helper function instead.

The duplication also made me realise I really want C++23's "deducing this" in Cpp2, so I've created https://github.com/hsutter/cppfront/issues/1197 for that.